### PR TITLE
Raft health + repair

### PR DESCRIPTION
**Pull request by samuelWilliams99**
Adds ability to damage + destroy raft pieces
Adds tape item, which heals raft parts in exchange for durability

image

TODO:
- [x] Tape model
- [x] Handle orphan raft parts

### CODE REVIEW

File: entities/weapons/rvr_repair_tool/cl_init.lua [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    local hp = aimEnt:Health()
+    local maxHp = aimEnt:GetMaxHealth()
+
+    local prog = hp / maxHp
+    col = HSVToColor( prog * 120, 1, 1 )
+
+    halo.Add( { aimEnt }, col, 5, 5, 2 )
+end
+
+function SWEP:GetHudIdentifier()
```
Comment (plally) [inline entities/weapons/rvr_repair_tool/cl_init.lua]: a  better name would be `GetHookIdentifier`. it is used for halos as well

File: gamemode/modules/rafts/sh_raft_class.lua [hunk 2]
```diff
@@ -83,6 +80,8 @@ function raftMeta:GetSpawnPosition( ply )
     local _, highestPiece = next( self.pieces )
 
     for _, raftPiece in pairs( self.pieces ) do
+        if not IsValid( raftPiece ) then continue end
```
Comment (plally) [inline gamemode/modules/rafts/sh_raft_class.lua]: perhaps you could put alot of this in a local function then the loop would become something like  
```lua
for _, raftPiece in pairs( self.pieces ) do
    if IsValid( raftPiece ) then
        local pos = getValidSpawnPos( ply, raftPiece )
        if pos then return pos end

        if raftPiece:GetPos().z > highestPiece:GetPos().z then
            highestPiece = raftPiece
        end
    end
end
```
(im not saying I require this, if you like the continue thats fine)

**Review by plally — CHANGES_REQUESTED** [review]

File: gamemode/modules/rafts/sh_raft_class.lua [hunk 3]
```diff
@@ -75,34 +72,41 @@ function raftMeta:GetPosition( piece )
     return piece:GetRaftGridPosition()
 end
 
+local config = ( GAMEMODE or GM ).Config.Rafts
+
+local function getValidSpawnPos( ply, raftPiece, plyMins, plyMaxs )
+    if not IsValid( raftPiece ) then return end
+
+    if config.SPAWNPOINT_PARTS[raftPiece:GetClass()] then
```
Comment (plally) [inline gamemode/modules/rafts/sh_raft_class.lua]: return early

File: gamemode/modules/rafts/sh_raft_class.lua [hunk 4]
```diff
@@ -137,6 +141,106 @@ function raftMeta:CanBuild( ply )
...
+        return
+    end
+
+    if piece:GetClass() ~= "raft_foundation" then return end
+
+    if piece.removedByRaft then return end
+
+    local grid, piecesGrid = self:GetSegmentingGrid()
+
+    local segments = RVR.Util.segmentGrid( grid )
```
Comment (overlordbleck) [inline gamemode/modules/rafts/sh_raft_class.lua]:
```suggestion
    if piece:GetClass() ~= "raft_foundation" then return end
    if piece.removedByRaft then return end

    local grid, piecesGrid = self:GetSegmentingGrid()
    local segments = RVR.Util.segmentGrid( grid )
```

**Review by overlordbleck — COMMENTED** [review]
Better than PCL's

**Pull request merged**